lib/rust/ensogl/core/src/display/object/event.rs

```diff
@@ -0,0 +1,175 @@
+//! Events implementation. Events behave in a similar way to JavaScript Events. When an event is
+//! emitted, it is propagated in three stages: capturing, target, and bubbling. Each stage is
+//! configurable and some events propagation can be cancelled. To learn more about the mechanics,
+//! see: https://javascript.info/bubbling-and-capturing.
+
+use crate::prelude::*;
+
+use super::class::Instance;
+use super::class::WeakInstance;
+
+
+
+// =============
+// === State ===
+// =============
+
+/// Event state. It can be used to determine whether the event is being propagated, its propagation
+/// is cancelled, or that the propagation cannot be cancelled. See docs of this module to learn
+/// more.
+#[allow(missing_docs)]
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+pub enum State {
+    #[default]
+    Running,
+    RunningNonCancellable,
+    Cancelled,
+}
+
+
+
+// =================
+// === SomeEvent ===
+// =================
+
+/// A counterpart of [`Event`] with hidden payload type. It is used to construct, configure, and
+/// emit new events.
+#[allow(missing_docs)]
+#[derive(Clone, CloneRef, Debug)]
+pub struct SomeEvent {
+    pub data:     frp::AnyData,
+    state:        Rc<Cell<State>>,
+    pub captures: Rc<Cell<bool>>,
+    pub bubbles:  Rc<Cell<bool>>,
+}
+
+impl SomeEvent {
+    /// Constructor.
+    pub fn new<Host: 'static, T: 'static>(target: Option<&Instance<Host>>, payload: T) -> Self {
+        let event = Event::new(target.map(|t| t.downgrade()), payload);
+        let state = event.state.clone_ref();
+        let captures = Rc::new(Cell::new(true));
+        let bubbles = Rc::new(Cell::new(true));
+        Self { data: frp::AnyData::new(event), state, captures, bubbles }
+    }
+
+    /// The [`State]` of the event.
+    pub fn state(&self) -> State {
+        self.state.get()
+    }
+}
+
+impl Default for SomeEvent {
+    fn default() -> Self {
+        Self::new::<(), ()>(None, ())
+    }
+}
+
+
+
+// =============
+// === Event ===
+// =============
+
+/// The [`Event`] interface represents an event which takes place in the EnsoGL display object
+/// hierarchy.
+///
+/// An event can be triggered by the user action e.g. clicking the mouse button or tapping keyboard,
+/// or generated by APIs to represent the progress of an asynchronous task. It can also be triggered
+/// programmatically, such as by calling the [`display::object::Instance::focus()`] method of an
+/// element, or by defining the event, then sending it to a specified target using
+/// [`display::object::Instance::event_source::emit(...)`].
+///
+/// See: https://developer.mozilla.org/en-US/docs/Web/API/Event.
```

Comment:
Why I should see it? Is it used here? Is it exactly the same mechanism? Or similar? If only similar, then what are the differences?

Reply:
I changed it to `See the JavaScript counterpart of this struct:`. I hope it clarifies the intention.